internal/cache/store/queue.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2019 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"hash/fnv"
+	"sync"
+)
+
+// ShardedUniqueQueue is a queue of write requests
+// for objects. It compacts consecutive create and update
+// requests, provides a slice of channels for consumers.
+// Requests consumed from these channels are mutually
+// exclusive
+type ShardedUniqueQueue interface {
+	AddIfAbsent(Request)
+	GetConsumers() []<-chan func() Request
+}
+
+// NewShardedUniqueQueue creates a sharded queue of write requests
+func NewShardedUniqueQueue(buckets int) ShardedUniqueQueue {
+	queues := make([]chan func() Request, 0, buckets)
+	for i := 0; i < buckets; i++ {
+		queues = append(queues, make(chan func() Request, 100))
```

Comment:
What is 100, can we extract that magic number somewhere?

Reply:
We can, it is the number of in flight write requests we can tolerate until we start blocking for additional write requests